README.md
# Upcoming changes
--------------------------------------------------------------------------------
  * In the function determineNextEvent () the last target time will be the
Period length added to calculate the next execution time.
If the connection was interrupted for a long time, however, they accumulate
many versions of the then with a shorter period length
(tv_earliest_next_trigger_time) are processed. If the cycle times anyway
are already tightly laid out, this can lead to an overload. Must
be changed. The next execution time must be the current time plus
Be cycle time.

--------------------------------------------------------------------------------

# Error Codes in Process Image
## ModbusTCP

| Modbus_Master_Status ||                                              |
|----|:---------------:|-----------------------------------------------|
| 16 | 0x10            | Initialization is failed                      |
| 17 | 0x11            | This device is not readable(false IP, Cabel?) |


| Modbus_Action_Status|    |                                                                                                                          |
|---|----------------------|--------------------------------------------------------------------------------------------------------------------------|
|  1| ILLEGAL FUNCTION     | The applied function code is not allowed.<br/> Check whether the right function code is used.                            |
|  2| ILLEGAL DATA ADDRESS | The applied address is not valid. The register is either read-only or not existed.<br/> Please check the address.        |
|  3| ILLEGAL DATA VALUE   | At least one part of the applied data is invalid. It is possible that the count of register is exceed.<br/> Please check the value.|
| 13| INVALID DATA         | The Slave answered with an unknown package to me. This can occur, for example, after the connection has been interrupted.<br/> Please check your wiring.|
|104| CONNECTION RESET BY<br> PEER  | This is a TCP problem. It can occur if the addressed host has no instance of the modbus slave running.          |
|110| CONNECTION TIMED OUT | The slave didn't respond in time or not at all.<br/> Please check your configuration and cabling.                        |


## ModbusRTU

| Modbus_Master_Status ||                                                      |
|----|:---------------:|-------------------------------------------------------|
| 16 | 0x10            | Initialization is failed                              |
| 17 | 0x11            | Serial connection can't be opened (wrong device name, invalid configuration) |


| Modbus_Action_Status |   |                                                                                                                          |
|---|----------------------|--------------------------------------------------------------------------------------------------------------------------|
|  1| ILLEGAL FUNCTION     | The applied function code is not allowed.<br/> Check whether the right function code is used.                            |
|  2| ILLEGAL DATA ADDRESS | The applied address is not valid. The register is either read-only or not existed.<br/> Please check the address.        |
|  3| ILLEGAL DATA VALUE   | At least one part of the applied data is invalid. It is possible that the count of register is exceed.<br/> Please check the value.|
| 11| FAILED TO RESPOND    | Resource temporarily unavailable. Usually means that the device is not present on the network.                           |
| 12| INVALID CRC          | A disturbed packet was received from the slave. This can occur, for example, after the connection has been interrupted.<br/> Please check your wiring.|
| 13| INVALID DATA         | An incomplete packet was received from the slave. This can occur, for example, after the connection has been interrupted.<br/> Please check your wiring.|
|110| CONNECTION TIMED OUT | The slave didn't respond in time or not at all.<br/> Please check your configuration and cabling.                        |


## All internal error messages that do not occur and do not need to be documented
| ID|                                   |
|---|-----------------------------------|
|1  |  Illegal function                 |
|2  |  Illegal data address             |
|3  |  Illegal data value               |
|4  |  Slave device or server failure   |
|5  |  Acknowledge                      |
|6  |  Slave device or server is busy   |
|7  |  Negative acknowledge             |
|8  |  Memory parity error              |
|10 |  Gateway path unavailable         |
|11 |  Target device failed to respond  |
|12 |  Invalid CRC                      |
|13 |  Invalid data                     |
|14 |  Invalid exception code           |
|15 |  Too many data                    |


# Example configuration for the config.rsc

Further information in document [about configuration and I/O](doc/io.md)

```json
{

    "config": {
        "modbusActions": [
            {
                "SlaveAddress" : 1,
                # Address of the Modbus slave which is addressed by the
                # command

                "FunctionCode" : 2,
                # Modbus function code see below

                "RegisterAddress" : 1,
                # Modbus register address (start address) of the slave

                "QuantityOfRegisters" : 8,
                # Number of Modbus registers that are written / read

                "ActionInterval" : 100000,
                # Time interval in which the command is sent from the
                # master in µsec. For the user, information in 1 ms
                # steps is sufficient

                "ProcessImageStartByte" : 2,
                # The offset in the Pi process image for the Modbus data

                "Action ID" : 1
                # Ascending ID for the command for information in the
                # event of an error
            },
            {
                "SlaveAddress" : 1,
                "FunctionCode" : 3,
                "RegisterAddress" : 1200,
                "QuantityOfRegisters" : 8,
                "ActionInterval" : 200000,
                "ProcessImageStartByte" : 18,
                "Action ID" : 2
            },
            {
                "SlaveAddress" : 2,
                "FunctionCode" : 16,
                "RegisterAddress" : 1024,
                "QuantityOfRegisters" : 8,
                "ActionInterval" : 500000,
                "ProcessImageStartByte" : 2,
                "Action ID" : 3
            }
        ]
    }
}
```


# Modbus Function Codes:

Important function codes are marked with '*'.The other function codes may not
need to be supported.

| | Code | ID |
|-|----|------|
|*|    eREAD_COILS               | 0x01|
|*|    eREAD_DISCRETE_INPUTS     | 0x02|
|*|    eREAD_HOLDING_REGISTERS   | 0x03|
|*|    eREAD_INPUT_REGISTERS     | 0x04|
|*|    eWRITE_SINGLE_COIL        | 0x05|
|*|    eWRITE_SINGLE_REGISTER    | 0x06|
| |    eREAD_EXCEPTION_STATUS    | 0x07|
|*|    eWRITE_MULTIPLE_COILS     | 0x0F|
|*|    eWRITE_MULTIPLE_REGISTERS | 0x10|
| |    eREPORT_SLAVE_ID          | 0x11|
| |    eWRITE_MASK_REGISTER      | 0x16|
| |    eWRITE_AND_READ_REGISTERS | 0x17|

Coils and Discrete Inputs are one bit in size. In the process image, however,
one byte should be reserved for each bit.Holding registers and input registers
have a size of 2 bytes.

For each Modbus command, an additional byte must be reserved in the Pi process
image for error messages.

For the virtual device, a byte must also be reserved in the Pi process image for
error messages.

For each Modbus command and the virtual device, a bit must be reserved to reset
the error byte in the Pi process image


ChangeLog.txt
piModbus Master und Slave Changelog
=============================================================




Anstehende Änderungen
--------------------------------------------------------------
* in der Funktion determineNextEvent() wird wird zu letzten Soll-Zeit die 
Periodenlänge addiert um die den nächsten Ausführungszeitpunkt zu berchnen. 
Wenn die Verbindung aber längere Zeit unterbrochen war sammeln sich dadurch 
viele Ausführungen an die dann mit einer kürzeren Periodenlänge 
(tv_earliest_next_trigger_time) abgearbeitet werden. Wenn die Zykluszeiten eh 
schon knapp ausgelegt sind, kann das zu einer Überlastung führen. Das muss 
geändert werden. Die nächste Ausführungszeit muss die akutelle Zeit plus 
Zykluszeit sein.




src/modbusconfig.h
#pragma once

#include <sys/types.h>
#include <stdint.h>
#include <modbus/modbus.h>
#include <limits.h>
#include <arpa/inet.h>
#include <sys/queue.h>


#define MODBUS_RTU_MIN_DATABITS 5
#define MODBUS_RTU_MAX_DATABITS 8
#define MODBUS_RTU_MIN_STOPBITS 1
#define MODBUS_RTU_MAX_STOPBITS 2



//max registersize/bitsize for modbus functions
#define MAX_MODBUS_READ_COILS_COUNT             2000
#define MAX_MODBUS_READ_DISCRETE_INPUTS_COUNT   2000
#define MAX_MODBUS_READ_HOLDING_REGISTERS_COUNT 125
#define MAX_MODBUS_READ_INPUT_REGISTERS_COUNT   125
#define MAX_MODBUS_WRITE_COILS_COUNT            1968
#define MAX_MODBUS_WRITE_REGISTERS_COUNT        123

typedef enum
{
    eProtRTU,
    eProtTCP,
} EModbusProtocol;

typedef struct
{
    char szTcpIpAddress[INET_ADDRSTRLEN];     // String in numbers-and-dots notation ("a.b.c.d")
    int32_t i32uPort;
    int32_t maxModbusConnections;
} TTcpConfig;

typedef enum
{
    eREAD_COILS               = 0x01,
    eREAD_DISCRETE_INPUTS     = 0x02,
    eREAD_HOLDING_REGISTERS   = 0x03,
    eREAD_INPUT_REGISTERS     = 0x04,
    eWRITE_SINGLE_COIL        = 0x05,
    eWRITE_SINGLE_REGISTER    = 0x06,
    eREAD_EXCEPTION_STATUS    = 0x07,	//serial line only, not supported by libmodbus v3.0.6
    eWRITE_MULTIPLE_COILS     = 0x0F,
    eWRITE_MULTIPLE_REGISTERS = 0x10,
    eREPORT_SLAVE_ID          = 0x11,
    eWRITE_MASK_REGISTER      = 0x16,
    eWRITE_AND_READ_REGISTERS = 0x17,
} EModbusFunction;

typedef enum
{
    eNoError                = 0x00,     // no error
    eNoDevice               = 0x10,
    eNoResponseFromDevice   = 0x11,
    eModbusActionBacklog    = 0x12,
    eInternalError          = 0xf0,
} EModbusErrors;

typedef struct  
{
    uint16_t i16uVendorId;
    uint16_t i16uModelId;
    char astrSerialNumber[256];
    uint32_t i32uBaud;
    char cParity;					// Could be 'N', 'E' or 'O'
    uint8_t i8uDatabits;			// Could be 5, 6, 7 or 8
    uint8_t i8uStopbits;			// Could be 1 or 2
    uint8_t u8DeviceModbusAddress;	//Modbus device addresse between 1 and 255, 0 is reserved for broadcast
    char sz8DeviceFilePath[PATH_MAX];	//path to the specified device file e.g. "/dev/ttyUSB0"
} TRtuConfig;

typedef struct
{
    uint16_t i16uActionID;
    uint8_t i8uSlaveAddress;			//modbus slave address from 0(broadcast) to 
    EModbusFunction eFunctionCode;
    uint32_t i32uStartRegister;			//modbus adresses from 0x1 to 0x10000
    uint16_t i16uRegisterCount;			//data length in bits, (bytes) or words, depends on modbus function 
    uint32_t i32uInterval_us;			//interval in micro seconds
    uint32_t i32uStartByteProcessData;
    uint8_t i8uStartBitProcessData;
    uint32_t i32uStatusByteProcessImageOffset;		//The pi process image offset for the commands status byte
    uint32_t i32uResetStatusProcessImageByteOffset;	//The pi process image byte offset for the status reset
    uint8_t	i8uResetStatusProcessImageBitOffset;	//The pi process image bit offset for the status reset
        
} TModbusAction;

struct TMBActionEntry
{
    TModbusAction modbusAction;
    SLIST_ENTRY(TMBActionEntry) entries;
};


typedef struct
{
    uint16_t u16DiscreteInputs;		// # of discrete inputs
    uint16_t u16Coils;				// # of coils
    uint16_t u16InputRegisters;		// # of input registers
    uint16_t u16HoldingRegisters;	// # of holding registers
}TModbusSlaveDataSizeConfig;

typedef struct
{
    uint32_t u32DiscreteInputsOffset;		    // process image offset in bytes
    uint32_t u32DiscreteInputsLength;		    // data length in bits
    uint32_t u32CoilsInputOffset;			    // process image offset in bytes
    uint32_t u32CoilsLength;				    // data length in bits
    uint32_t u32InputRegistersOffset;		    // process image offset in bytes
    uint32_t u32InputRegistersLength;		    // data length in bytes
    uint32_t u32HoldingRegistersInputOffset;	// process image offset in bytes
    uint32_t u32HoldingRegistersLength;			// data length in bytes
}TProcessImageConfiguration;

typedef struct
{
    EModbusProtocol eProtocol;       // TCP or RTU
    union 
    {
        TTcpConfig tTcpConfig;
        TRtuConfig tRtuConfig;
    } uProt;
    int32_t i32uDeviceStatusByteProcessImageOffset;             //the device status byte offset in the pi process image
    int32_t i32uDeviceStatusResetByteProcessImageByteOffset;    //status reset byte offset
} TModbusDeviceConfiguration;



typedef struct
{
    TModbusDeviceConfiguration tModbusDeviceConfig;
    int32_t i32ActionCount;
    SLIST_HEAD(TMBActionListHead, TMBActionEntry) mbActionListHead; // Array of demanded actions, last element must be NULL
} TModbusMasterConfiguration;

typedef struct
{
    TModbusDeviceConfiguration tModbusDeviceConfig;
    TModbusSlaveDataSizeConfig tModbusDataConfig;
    TProcessImageConfiguration tProcessImageConfig;
} TModbusSlaveConfiguration;

struct TMBSlaveConfigEntry
{
    TModbusSlaveConfiguration mbSlaveConfig;
    SLIST_ENTRY(TMBSlaveConfigEntry) entries;
};

struct TMBMasterConfigEntry
{
    TModbusMasterConfiguration mbMasterConfig;
    SLIST_ENTRY(TMBMasterConfigEntry) entries;
};

SLIST_HEAD(TMBSlaveConfHead, TMBSlaveConfigEntry) mbSlaveConfHead;
SLIST_HEAD(TMBMasterConfHead, TMBMasterConfigEntry) mbMasterConfHead;



